### ISSUE TITLE
Insights on gstore-node

### ISSUE THREAD

**Issue by sebelga** [sebelga/gstore-node#101]
Hello to all of you using gstore-node!

I'd like to ask you something... What do you think could be done better to make gstore-node more visible/used by Node.js developers that are currently _simply_ using the "google-cloud/datastore" package?

When I see the stats in npm of both packages:
- google-cloud/datastore: **23,304** downloads in the last 7 days
- gstore-node: **400** downloads in the last 7 days

As gstore-node offers exactly the same API + many more features, I do wonder what is the reason for such a difference... Maybe the Datastore is not used/thought with Schemas in mind?

Happy to hear some feedback! 😄

**Comment by pauloddr** [comment]
Datastore schemas are so dynamic, down to the point of defining indexes per record/entity, as opposed to indexes on the entire Table/Kind. So I can create one record with an index in one column, and another record without that index in the same column. So you might be onto something about Datastores not being used with schemas in mind. I see the usefulness in schema validation though.

One thing I'm a little concerned about is memory consumption. I made a simple Google Function using gstore-node and a single request to the API used 80MB of memory. I think it's a bit too much, but I'm not sure whether this is coming from gstore-node or google/datastore, or my own code. For comparison, a Google Function using Sequelize connecting to MySQL uses about 30MB or memory. I didn't have the chance to test things yet as I'm not currently working on APIs and it's going to take a while until I can. But once I have a chance, I'll let you know of my findings.

**Comment by trollkotze** [comment]
@sebelga A use case for not wanting to index a property for some entities might be properties that can be larger than 1500 bytes on occasion.
Datastore would not be able to index those and throw an error if one has not specified that property to not be indexed for that entity. (But one might want it to be indexed for most other entities anyways.)
But I guess it would still be possible with `gstore` to set `excludeFromIndexes` on a per-entity basis? I have not yet probed deeper here. But this lib looks nice!

Does this library offer any kind of functionality to easily keep track of entity counts by any chance? This is a problem I am stuck with at the moment (using sharding counters but running into problems with transactions expiring too quickly under heavy load with lots of updates).

I would really like to try this out. Hope I'll find the time. :) (If this lib would help me with the counting problem, I think that would save me a lot of time.)

**Comment by sebelga** [comment]
@trollkotze 
For the indexes, yes it's possible to excludeFromIndexes for a property but (for now) it "sticks". That means that the next entity with the same schema would have that property excludeFromIndexes set to true.
So, to understand you well, you calculate at _runtime_ the size of a string, and according to the size of the string you decide to exclude it or not from indexes?  But then I must ask, why do you index it to begin with?

For your "counts" issue, I have in my todo list (for some time now) the task to add an abstraction to automatically shard entities with a counter built-in :) But I haven't had the chance to work on it yet. Can I ask what do you mean by "transactions expiring too quickly"?
For now, one solution I see is using a Redis db for the counter.  With gstore-node it's very easy with it's cache (https://sebelga.gitbooks.io/gstore-node/content/cache.html).

```js
const gstore = require('gstore-node')();
const { cache } = gstore;

// read
cache.get('user-count').then(...);

// write
cache.set('user-count', 12).then(...);
```

**Comment by trollkotze** [comment]
@sebelga Thanks for getting back!

> Can I ask what do you mean by "transactions expiring too quickly"?

Maybe better to put it the other way around: The operations inside the counter increment transaction (that is not really more than something like: `get(count); count+=increment; save(count)` ) run too slowly. So the transaction expires (after 60 seconds, that is, if my understanding is correct).

My sharding counter at the moment does the following, parallel to saving `x` new entities of a kind:

  1. pick a random shard
  2. create and start a [transaction](https://cloud.google.com/datastore/docs/concepts/transactions), and inside that transaction do:
    a) get the current value of that shard.
    b) increment by x and save it again.
    c) commit the transaction
  3. On failure: rollback the transaction. <- (Even that would fail under heavy load, with the same "transaction expired" error, but I guess that should then be okay:? However in that case, I'd be wondering what the "rollback" actually method is for, if a failed transaction would not have had any effects anyway. :? It seems I don't really understand transactions.)

The error I am commonly getting under heavy load is:
> Error: 3 INVALID_ARGUMENT: The referenced transaction has expired or is no longer valid

If I understood correctly, the lifetime of a datastore transaction is 60 seconds. So that shows me that my simple getting and saving of a counter inside that transaction is really slow. But as I said, this happens only under heavy load. When doing a few single updates, the counting works.

(Maybe my problem is just my not so ultrahigh bandwidth internet connection in my local dev environment and it would be a total non-issue when hosted on AppEngine, where we want to deploy our app. But if a mere slowdown like this for a short period breaks crucial functionality, then I think that is unacceptable, so I'd better fix it.)

I retry the above with exponential backoff. (Yes, I'm creating a new transaction in every retry, not reusing the old txn object again and failing because of that.) When still failing, cache the unprocessed increment amount in memory, and let it be picked up and added to the next increment.

This is a really bad and unsafe workaround, because I don't know when that will happen. The app might crash or be restarted before that happens. And it could keep the count wrong for a long time.

Meanwhile, the simultaneously running "normal" (i.e. non-transactional) saves of the entities which I am counting succeed without problem. Only the transactional counter updates get stuck in this exponential backoff retry-fail-again loop.

But now when thinking about it: I should perhaps better not allow even trying to do multiple increments in parallel (which would then all retry, and mostly fail, with exponential backoff, slowing each other down and leading to more accumulating failures), but instead just turn it into a queue:

When a count increment by `x` should be done:
  1. Increment an in-memory `"unsynced"` counter for that entity kind by `x`.
  2. 
     a) If a count transaction for that entity kind is already running, return.
     b) Else, start a count transaction, trying to increment by the current `"unsynced"` amount.

When a count transaction finishes:

  1. a) successfully: decrement the in-memory `"unsynced"` counter by the amount that has been counted in the transaction.
     b) failed: leave the `"unsynced"` counter as it is
  2.
     If the in memory counter `"unsynced > 0"`, start a new count transaction, trying to increment by the current `"unsynced"` amount.

I had thought that, when I have multiple shards, it should not be a problem to do multiple increments on different shards at the same time. But this seems to be problematic when the number of parallel running counter increment transactions becomes larger and larger, and especially, I guess, when bandwidth is somewhat limited.

Thanks so much for asking and letting me think about it aloud here. Now I think I'm a bit clearer about what is going on. :smiley:


**Re: the cache solution,** with a connected redis server. Yay! That was exactly what I thought yesterday when I first saw this library. This looks seductively simple.

.

(Seems I have a bit hijacked this thread here now with my problems. Maybe this would in part fit better into a separate feature issue for a built-in sharding counter, but I can't come up with a clean enough lead-in to that right now.
Instead, - my apologies, - I still have some more of my particular problems to talk about just real quick, while I'm already at it. Would appreciate some comment, but won't bug any further about that afterwards, I promise. :flushed: )

.

Of course the redis cache solution would require additional infrastructure (a running redis server somewhere). And this seems possibly quite expensive? Since we are using AppEngine, I quickly found this "click-to-deploy solution" here of hosting a [redis server on Google Compute](https://console.cloud.google.com/launcher/details/click-to-deploy-images/redis):  and was blown away by the price. For just a few counters that we would need, this does not seem worth it. But maybe a much scaled-down version might do and be affordable. I have no idea about scaling such things.
I have even seen a [redis-server](https://www.npmjs.com/package/redis-server) package for Node.js. Do you think that might be feasible to run on AppEngine in a Node.js instance that is always online and already doing some other stuff?

**Comment by sebelga** [comment]
Hi @trollkotze 
Sorry for coming back late on this. Do you mind creating a new issue with this as a question so this thread does not change the focus?

You are probably doing something wrong as 60 seconds is a lot 😄 But it'd be better to get help and insight in a separate issue ticket.

Just for info https://redislabs.com/ has a free tier of 30MB which is plenty for a counter 👍 

Cheers!

**Comment by trollkotze** [comment]
Yes, I had that on mind, now especially after having looked deeper into your library, but deciding not to use it, because I see a lot to improve.

I have not much time at the moment to get deeper into this and provide constructive criticism or ideas, because I have to get finished with something. But it has helped me to find into Redis at least, which I am now using besides Cloud Datastore, and I hope to come back here when I find time, because I really like the idea of this project.

And yes, I am already using RedisLabs free tier now for testing, will use enterprise tier in production as well. Thanks for the tip!

(It seems I cannot delete my long rambling comment from here, after I just moved the issue [over to here](https://github.com/sebelga/gstore-node/issues/107). I guess it makes sense to keep this thread clean, so feel free to remove it here.)

**Comment by sebelga** [comment]
@trollkotze 
> "but deciding not to use it, because I see a lot to improve."

Looking forward for your insight and how we could improve the lib. Whenever you have time! :+1:

**Comment by carnun** [comment]
@sebelga I'm trying to remember how I first found your package... I think it was via `mongoose` and searching for something similar for Google Datastore. It was just a lucky find on my part, otherwise I would have gone the raw `@google-cloud/datastore` route. I suspect the download stats are more a visibility thing, potentially having other people blogging about using it and their use cases may change that.

I use Datastore as a backend for GraphQL APIs and `gstore-node` fits really nicely with my stack. One of the items on my todo list is to create a sequence of blogs that discuss my experiences with this stack so once I get round to doing that I'll ping you.

**Comment by sebelga** [comment]
Thanks for the input @carnun ! Yes, I do think the visibility of gstore might be improved, I wish I had more time to blog about it. Let me know about your blog posts when you have them, I am interested to know more about your stack (didn't know about TypeGraphQL, very interesting).

**Comment by carnun** [comment]
Closing this thread. New instances of Google Cloud Datastore can now support the Mongo API it is probably worthwhile to rather use `mongoose` for new projects.
